### PR TITLE
qa.md is done

### PR DESCRIPTION
**Pull request by AA1HSHH**
Maybe have some mistakes I will keep tracking.

### CODE REVIEW

File: _2020/qa.md [hunk 1]
```diff
@@ -1,182 +1,183 @@
...
+## 学习操作系统相关内容的推荐，比如进程，虚拟内存，中断，内存管理等
 
-- [MIT's 6.828 class](https://pdos.csail.mit.edu/6.828/) - Graduate level class on Operating System Engineering. Class materials are publicly available.
-- Modern Operating Systems (4th ed) - by Andrew S. Tanenbaum is a good overview of many of the mentioned concepts.
-- The Design and Implementation of the FreeBSD Operating System - A good resource about the FreeBSD OS (note that this is not Linux). 
-- Other guides like [Writing an OS in Rust](https://os.phil-opp.com/) where people implement a kernel step by step in various languages, mostly for teaching purposes. 
 
 
-## What are some of the tools you'd prioritize learning first?
+首先，不清楚你是不是真的需要熟悉这些 更底层的话题。
```
Comment (hanxiaomax) [inline _2020/qa.md:36]:
```suggestion
首先，不清楚你是不是真的需要熟悉这些更底层的话题。
```

Comment (hanxiaomax) [reply _2020/qa.md:36]: 还有多处有这种问题，自己检查一下吧

File: _2020/qa.md [hunk 2]
```diff
@@ -1,182 +1,183 @@
...
+学习资源：
 
-- Learning how to use your keyboard more and your mouse less. This can be through keyboard shortcuts, changing interfaces, &c.
-- Learning your editor well. As a programmer most of your time is spent editing files so it really pays off to learn this skill well.
-- Learning how to automate and/or simplify repetitive tasks in your workflow because the time savings will be enormous...
-- Learning about version control tools like Git and how to use it in conjunction with GitHub to collaborate in modern software projects.
+- [MIT's 6.828 class](https://pdos.csail.mit.edu/6.828/) -  研究生阶段的操作系统课程（课程资料是公开的）。
+- 现代操作系统 第四版（Modern Operating Systems 4th ed） -  作者是Andrew S. Tanenbaum 这本书对上述很多概念都有很好的描述。
+- FreeBSD的设计与实现（The Design and Implementation of the FreeBSD Operating System） - 关于FreeBSD OS 的好资源(注意，FreeBSD OS不是Linux)。
+- 其他的指南例如 [用 Rust写操作系统](https://os.phil-opp.com/) 这里用不同的语言 逐步实现了内核，主要用于教学的目的。
```
Comment (hanxiaomax) [inline _2020/qa.md:44]: 英文单词两侧要有空格，英文书名用斜体。全文记得一并检查一下

File: _2020/qa.md [hunk 3]
```diff
@@ -1,182 +1,183 @@
...
 
-## What are the advantages and disadvantages of each OS and how can we choose between them (e.g. choosing the best Linux distribution for our purposes)?
+## Docker和虚拟机有什么区别?
 
-Regarding Linux distros, even though there are many, many distros, most of them will behave fairly identically for most use cases. 
-Most of Linux and UNIX features and inner workings can be learned in any distro. 
-A fundamental difference between distros is how they deal with package updates. 
-Some distros, like Arch Linux, use a rolling update policy where things are bleeding-edge but things might break every so often. On the other hand, some distros like Debian, CentOS or Ubuntu LTS releases are much more conservative with releasing updates in their repositories so things are usually more stable at the expense of sacrificing newer features. 
-Our recommendation for an easy and stable experience with both desktops and servers is to use Debian or Ubuntu.
+Docker 基于更加普遍的被称为容器的概念。关于容器和虚拟机之间最大的不同是 虚拟机会执行整个的 OS 栈，包括内核（即使这个内核和主机内核相同）。与虚拟机不同的是，容器避免运行其他内核实例 ，而是与主机分享内核。在Linux环境中，有LXC机制来实现，并且这能使一系列分离的主机像是在使用自己的硬件启动程序，而实际上是共享主机的硬件和内核。因此容器的开销小于完整的虚拟机。
```
Comment (hanxiaomax) [inline _2020/qa.md:134]:
```suggestion
Docker 基于容器这个更为概括性的概念。关于容器和虚拟机之间最大的不同是，虚拟机会执行完整的 OS 栈，包括内核（即使这个内核和主机内核相同）。与虚拟机不同的是，容器避免运行其他内核实例 ，而是与主机分享内核。在Linux环境中，有LXC机制来实现，并且这能使一系列分离的主机像是在使用自己的硬件启动程序，而实际上是共享主机的硬件和内核。因此容器的开销小于完整的虚拟机。
```

File: _2020/qa.md [hunk 4]
```diff
@@ -1,182 +1,183 @@
...
 
-Mac OS is a good middle point between Windows and Linux that has a nicely polished interface. However, Mac OS is based on BSD rather than Linux, so some parts of the system and commands are different.
-An alternative worth checking is FreeBSD. Even though some programs will not run on FreeBSD, the BSD ecosystem is much less fragmented and better documented than Linux. 
-We discourage Windows for anything but for developing Windows applications or if there is some deal breaker feature that you need, like good driver support for gaming. 
+另一方面，容器的隔离性较弱而且只有在主机运行相同的内核时才能正常工作。例如，如果你在macOS上运行Docker，Docker需要启动Linux虚拟机去获取初始的Linux内核，这样的开销仍然很大。最后，Docker是容器的特定实现，它是为软件部署定制的。基于这些，它有一些奇怪之处：例如，默认情况下，Docker容器在重启之间不会有以任何形式的存储。
 
-For dual boot systems, we think that the most working implementation is macOS' bootcamp and that any other combination can be problematic  on the long run, specially if you combine it with other features like disk encryption. 
+## 每种OS的优缺点是什么，我们如何选择（比如选择最适用于我们需求的Linux发行版)?
```
Comment (hanxiaomax) [inline _2020/qa.md:138]:
```suggestion
## 不同操作系统的优缺点是什么，我们该如何选择（比如选择最适用于我们需求的Linux发行版)?
```

File: _2020/qa.md [hunk 5]
```diff
@@ -1,182 +1,183 @@
...
-- Leader Key - The leader key is a special key that is often left to the user to be configured for custom commands. The pattern is usually to press and release this key (often the space key) and then some other key to execute a certain command. Often, plugins will use this key to add their own functionality, for instance the UndoTree plugin uses `<Leader> U` to open the undo tree. 
-- Advanced Text Objects - Text objects like searches can also be composed with vim commands. E.g. `d/<pattern>` will delete to the next match of said pattern or `cgn` will change the next occurrence of the last searched string. 
+课程的一些经验可以直接用于机器学习程序。
+就像许多科学学科一样，在机器学习中，你经常要进行一系列实验，并检查哪些数据有效，哪些无效。
+你可以使用Shell轻松快速地搜索这些实验结果，并且以合理的方式汇总。这意味着需要在限定时间内或使用特定数据集的情况下，检查所有实验结果。通过使用JSON文件记录实验的所有相关参数，使用我们在本课程中介绍的工具，这件事情可以变得极其简单。
+最后，如果你不使用集群提交你的GPU作业，那你应该研究如何使该过程自动化，因为这是一项非常耗时的任务，会消耗你的精力。
 
-## What is 2FA and why should I use it?
+## 还有更多的Vim提示吗？
```
Comment (hanxiaomax) [inline _2020/qa.md:163]:
```suggestion
## 还有更多的 Vim 小窍门吗？
```

File: _2020/qa.md [hunk 6]
```diff
@@ -1,182 +1,183 @@
...
-Another browser called [Flow](https://www.ekioh.com/flow-browser/) is not user ready yet, but it is implementing a new rendering engine that promises to be faster than the current ones. 
+## 2FA是什么，为什么我需要使用它?
+
+双因子验证（Two Factor Authentication  2FA）在密码之上为帐户增加了一层额外的保护。为了登录，你不仅需要知道密码，还必须以某种方式“证明”可以访问某些硬件设备。最简单的情形是可以通过接收手机的SMS来实现（尽管SMS 2FA 存在 [已知问题](https://www.kaspersky.com/blog/2fa-practical-guide/24219/)）。我们推荐使用[YubiKey](https://www.yubico.com/)之类的[U2F](https://en.wikipedia.org/wiki/Universal_2nd_Factor)方案。
+
+## 对于不同的Web浏览器有什么评价?
+
+2020的浏览器现状是，大部分的浏览器都与Chrome 类似，因为他们都使用同样的引擎(Blink)。 Microsoft Edge同样基于 Blink，至于Safari 基于 WebKit(与Blink类似的引擎)，这些浏览器仅仅是更糟糕的Chorme版本。不管是在性能还是可用性上，Chorme都是一款很不错的浏览器。如果你想要替代品，我们推荐Firefox。Firefox与Chorme的在各方面不相上下，并且在隐私方面更加出色。
```
Comment (hanxiaomax) [inline _2020/qa.md:182]:
```suggestion
2020的浏览器现状是，大部分的浏览器都与Chrome 类似，因为它们都使用同样的引擎(Blink)。 Microsoft Edge同样基于 Blink，至于Safari 基于 WebKit(与Blink类似的引擎)，这些浏览器仅仅是更糟糕的Chorme版本。不管是在性能还是可用性上，Chorme都是一款很不错的浏览器。如果你想要替代品，我们推荐Firefox。Firefox与Chorme的在各方面不相上下，并且在隐私方面更加出色。
```

File: _2020/qa.md [hunk 7]
```diff
@@ -1,182 +1,183 @@
...
-- Password Manager Integration - Most password managers have browser extensions that make inputting your credentials into websites not only more convenient but also more secure. Compared to simply copy-pasting your user and password, these tools will first check that the website domain matches the one listed for the entry, preventing phishing attacks that impersonate popular websites to steal credentials. 
 
-## What are other useful data wrangling tools?
+## 你使用那些浏览器插件?
 
-Some of the data wrangling tools we did not have time to cover during the data wrangling lecture include `jq` or `pup` which are specialized parsers for JSON and HTML data respectively. The Perl programming language is another good tool for more advanced data wrangling pipelines. Another trick is the `column -t` command that can be used to convert whitespace text (not necessarily aligned) into properly column aligned text.
+我们钟爱的插件主要与安全性与可用性有关：
```
Comment (hanxiaomax) [inline _2020/qa.md:117]:
```suggestion
我们钟爱的插件主要与安全性和可用性有关：
```

File: _2020/qa.md [hunk 8]
```diff
@@ -1,182 +1,183 @@
...
 
-- bash is easy to get right for a simple use case but it can be really hard to get right for all possible inputs. For example, spaces in script arguments have led to countless bugs in bash scripts.
-- bash is not amenable to code reuse so it can be hard to reuse components of previous programs you have written. More generally, there is no concept of software libraries in bash.
-- bash relies on many magic strings like `$?` or `$@` to refer to specific values, whereas other languages refer to them explicitly, like `exitCode` or `sys.args` respectively. 
+值得优先学习的内容：
 
-Therefore, for larger and/or more complex scripts we recommend using more mature scripting languages like Python or Ruby. 
-You can find online countless libraries that people have already written to solve common problems in these languages.
-If you find a library that implements the specific functionality you care about in some language, usually the best thing to do is to just use that language.
+- 学着更多去使用键盘，更少使用鼠标。可以用快捷键，更换界面等等。
```
Comment (hanxiaomax) [inline _2020/qa.md:51]:
```suggestion
- 学着多用键盘，少用鼠标。这一目标可以通过多加利用快捷键，更换界面等来实现
```

Comment (AA1HSHH) [reply _2020/qa.md]: 谢谢你，我来做一些修改。

**Comment by AA1HSHH** [comment]
@hanxiaomax  I have fixed some issues. Thanks for your suggestions.

**Comment by hanxiaomax** [comment]
👌

**Pull request merged**